build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/util/DependenciesUtils.java

```diff
@@ -47,4 +50,30 @@ public static FileCollection createFileCollectionFromNonTransitiveArtifactsView(
         }).getFiles();
     }
 
+    /**
+     * This method gives us an artifact view of a configuration that filters out all
+     * project dependencies that are not shadowed jars.
+     * Basically a thirdparty only view of the dependency tree.
+     */
+    public static FileCollection thirdPartyDependenciesView(Configuration configuration) {
```

Comment:
Doesn't calling this effectively trigger dependency resolution? I think we might be implicitly relying on lazy task configuration to not eagerly resolve dependencies when we shouldn't but I think we might need to wrap this in a `Provider` or something to be totally proper.

Reply:
The artifactview handling itself is lazy. But there was indeed an oversight in the util method implementation where I resolved the provider eagerly and the dependency infos (not the artifacts) have been resolved eagerly. fixed this here: https://github.com/elastic/elasticsearch/pull/118723/commits/5f507b04a9719e71a589ec21f78893020116fd69 

now those configurations are really only resolved at task graph calculation time. 

One issue is that I have configuration cache enabled by default for testing locally and this makes dependency resolution timings differ from the _normal way_ that makes it easier to oversee this 🤔

Reply:
But doesn't the call to `getFiles()` immediately trigger resolution? Where's the lazy bit?

Reply:
@mark-vieira getFiles only gives you back file collection that is lazy as configuration itself is. the resolution is kicking only in when that file collection is resolved. like a configuration is implementing  filecollection too. you might mix that up with FileCollection#getFiles like Configuration#getFiles returning a Set<File>. 

Trust me bro...

And in case you don't .... 
I double checked manually by throwing in exception using the beforeResolve hook and that shows how only the task graph building triggered the resolution: https://gradle-enterprise.elastic.co/s/n6igninpsi3m4/failure?expanded-stacktrace=WyIwIl0&focused-stack-frame=0-28#1

Reply:
I need to think if there is a way we can enforce failing builds or at least have a regular check running to ensure we don't resolve anything too early. One option might be to verify https://gradle-enterprise.elastic.co/s/6bqzp2o7mq6k2/performance/dependency-resolution#dependency-resolution-configuration-user-initiated programatically somehow.

Reply:
> One option might be to verify https://gradle-enterprise.elastic.co/s/6bqzp2o7mq6k2/performance/dependency-resolution#dependency-resolution-configuration-user-initiated programatically somehow.

Indeed. I wonder if that's detected during the built itself, or just inferred based on resolution event data after the fact.